code/iaas/allocator/src/main/java/io/cattle/platform/allocator/constraint/DiskSizeConstraint.java

```diff
@@ -0,0 +1,96 @@
+package io.cattle.platform.allocator.constraint;
+
+import io.cattle.platform.allocator.service.AllocationAttempt;
+import io.cattle.platform.allocator.service.AllocationCandidate;
+import io.cattle.platform.allocator.service.CacheManager;
+import io.cattle.platform.allocator.service.DiskInfo;
+import io.cattle.platform.allocator.service.HostInfo;
+import io.cattle.platform.allocator.service.InstanceDiskReserveInfo;
+import io.cattle.platform.allocator.service.InstanceInfo;
+import io.cattle.platform.object.ObjectManager;
+
+import java.util.Map.Entry;
+import java.util.Set;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class DiskSizeConstraint extends HardConstraint implements Constraint {
+
+    private static final Logger log = LoggerFactory.getLogger(DiskSizeConstraint.class);
+
+    private Long reserveSize;
+    private String volumeName;
+    private ObjectManager objectManager;
+
+    public DiskSizeConstraint(String volumeName, String sizeWithUnit, ObjectManager objMgr) {
+        this.volumeName = volumeName;
+        this.reserveSize = Long.parseLong(sizeWithUnit.replaceAll("[^0-9]", ""));
+        this.objectManager = objMgr;
+    }
+
+    @Override
+    public boolean matches(AllocationAttempt attempt, AllocationCandidate candidate) {
+        Set<Long> hostIds = candidate.getHosts();
+
+        // if one of the host does not have enough free space then return false
+        for (Long hostId : hostIds) {
+            boolean oneGood = false;
+
+            CacheManager cm = CacheManager.getCacheManagerInstance(this.objectManager);
+            HostInfo hostInfo = cm.getHostInfo(hostId);
+
+            // we will get a bunch of disks for that host and we need at least one disk with
+            // free space large enough
+            oneGood = checkDiskSize(cm, hostInfo, attempt.getInstanceId());
+
+            // if no disk with big enough free space for this host, then
+            // candidate is no good
+            if (!oneGood) {
+                log.debug("Scheduling instance [{}] to host [{}] rejected", attempt.getInstanceId(), hostId);
+                return false;
+            }
+        }
+        return true;
+    }
+
+    public boolean checkDiskSize (CacheManager cm, HostInfo hostInfo, Long instanceId) {
+
+        for (Entry<String, DiskInfo> entry : hostInfo.getAllDiskInfo()) {
+            DiskInfo diskInfo = entry.getValue();
+            Long capacity = entry.getValue().getCapacity();
```

Comment:
nit pick: you can use the diskInfo object instead of evaluating entry.getValue() again